examples/pytorch/distributed_stl10/stl10_example.py

```diff
@@ -0,0 +1,360 @@
+            batch_size = input[0].size(0)
+            losses.update(to_python_float(reduced_loss), batch_size)
+            top1.update(to_python_float(prec1), batch_size)
+            top5.update(to_python_float(prec5), batch_size)
+
+            torch.cuda.synchronize()
+            batch_time.update((time.time() - end) / log_freq)
+            end = time.time()
+
+            if i % log_freq == 0 and is_master:
+                # Only the first node should log infos.
+                print(
+                    f"Test: [{i}/{len(val_loader)}]\t"
+                    f"Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t"
+                    f"Speed {world_size * batch_size / batch_time.val:.3f} ({world_size * batch_size / batch_time.avg:.3f})\t"
+                    f"Loss {losses.val:.10f} ({losses.avg:.4f})\t"
+                    f"Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t"
+                    f"Prec@5 {top5.val:.3f} ({top5.avg:.3f})"
+                )
+
+        if is_master:
+            print(f' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}')
+
+    return top1.avg
+
+
+def run(batch_size: int,
+        epochs: int,
+        learning_rate: float,
+        log_interval: int,
+        save_model: bool):
+    # number of nodes / node ID
+    n_nodes = int(os.environ['SLURM_JOB_NUM_NODES'])
+    node_id = int(os.environ['SLURM_NODEID'])
+
+    # local rank on the current node / global rank
+    local_rank = int(os.environ['SLURM_LOCALID'])
+    global_rank = int(os.environ['SLURM_PROCID'])
+
+    # number of processes / GPUs per node
+    world_size = int(os.environ['SLURM_NTASKS'])
+    n_gpu_per_node = world_size // n_nodes
+
+    # define master address and master port
+    hostnames = subprocess.check_output(['scontrol', 'show', 'hostnames', os.environ['SLURM_JOB_NODELIST']])
+    master_addr = hostnames.split()[0].decode('utf-8')
+
+    # set environment variables for 'env://'
+    os.environ['MASTER_ADDR'] = master_addr
+    os.environ['MASTER_PORT'] = str(29500)
+    os.environ['WORLD_SIZE'] = str(world_size)
+    os.environ['RANK'] = str(global_rank)
+
+    # define whether this is the master process / if we are in distributed mode
+    is_master = node_id == 0 and local_rank == 0
+    multi_node = n_nodes > 1
+    multi_gpu = world_size > 1
+
+    # summary
+    PREFIX = "%i - " % global_rank
+    print(PREFIX + "Number of nodes: %i" % n_nodes)
+    print(PREFIX + "Node ID        : %i" % node_id)
+    print(PREFIX + "Local rank     : %i" % local_rank)
+    print(PREFIX + "Global rank    : %i" % global_rank)
+    print(PREFIX + "World size     : %i" % world_size)
+    print(PREFIX + "GPUs per node  : %i" % n_gpu_per_node)
+    print(PREFIX + "Master         : %s" % str(is_master))
+    print(PREFIX + "Multi-node     : %s" % str(multi_node))
+    print(PREFIX + "Multi-GPU      : %s" % str(multi_gpu))
+    print(PREFIX + "Hostname       : %s" % socket.gethostname())
+
+    # set GPU device
+    torch.cuda.set_device(local_rank)
```

Comment:
This can be problematic when training with a number of gpus != number of gpus of the nodes. An option is for example
`torch.cuda.set_device(int(os.environ['CUDA_VISIBLE_DEVICES'].split(',')[local_rank]))`

Reply:
I don't think this is the case. The Slurm Generic Resource (GRES) Scheduling should handle this by setting the `CUDA_VISIBLE_DEVICES` for each local task according to the definition inside the `pytorch_distributed_stl10_example.sh` script:

```bash
#SBATCH --ntasks=2                         
#SBATCH --ntasks-per-node=1         
#SBATCH --gres=gpu:1                      # sets CUDA_VISIBLE_DEVICES
```

Reply:
Oh, if CUDA_VISIBLE_DEVICES=3,4, torch.cuda.set_device(0) sets torch to use CUDA device 3 ? I always set CUDA_VISIBLE_DEVICES to be unique for each process but it is good to know !

Reply:
Yes! This is very useful in distributed.